### ISSUE TITLE
PyRight rejects function with `TypeVar` parameterized generic type and arguments referencing the same `TypeVar` variables

### ISSUE THREAD

**Issue by wjakob** [microsoft/pyright#7355]
**Describe the bug**
Pyright exhibits unexpected behavior when type-checking a function containing a TypeVar-parameterized generic type and other arguments containing those TypeVars, specifically when they occur within a union. The order within the union plays a role, as well. The snippet below demonstrates the issue.

The example demonstrates the use of a generic type `MyType` parameterized by itself, and two type variables `T1` and `T2`. Two functions `f1` and `f2` can be applied to `MyType` and an argument of `T1 | T2` (or `T2 | T1`, which should make no difference).

MyPy type-checks this code without errors or warnings.

**Code or Screenshots**
```python
from typing import TypeVar, Generic, cast, Union

Self = TypeVar("Self")
T1 = TypeVar("T1")
T2 = TypeVar("T2")

class MyType(Generic[Self, T1, T2]):
    pass

def f1(arg0: MyType[Self, T1, T2], arg1: T1 | T2, /) -> Self:
    return cast(Self, arg0)

def f2(arg0: MyType[Self, T1, T2], arg1: T2 | T1, /) -> Self:
    return cast(Self, arg0)

class IntStrMyType(MyType['IntStrMyType', int, str]):
    pass

v0 = IntStrMyType()
v1 = f1(v0, 1)        # fails
v1 = f1(v0, 'a')      # OK

v1 = f2(v0, 1)        # OK
v1 = f2(v0, 'a')      # fails
```

This fails with
```
/Users/wjakob/test.py
  /Users/wjakob/test.py:20:9 - error: Argument of type "IntStrMyType" cannot be assigned to parameter "arg0" of type "MyType[Self@f1, T1@f1, T2@f1]" in function "f1"
    "IntStrMyType" is incompatible with "MyType[IntStrMyType, int, str | int]"
      Type parameter "T2@MyType" is invariant, but "str" is not the same as "str | int" (reportArgumentType)
  /Users/wjakob/test.py:24:9 - error: Argument of type "IntStrMyType" cannot be assigned to parameter "arg0" of type "MyType[Self@f2, T1@f2, T2@f2]" in function "f2"
    "IntStrMyType" is incompatible with "MyType[IntStrMyType, int | str, str]"
      Type parameter "T1@MyType" is invariant, but "int" is not the same as "int | str" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

**VS Code extension or command-line**
I am using PyRight 1.1.351 on the command line.

**Comment by erictraut** [comment]
This issue is closely related to at least one other open issue: https://github.com/microsoft/pyright/issues/6957. 

The problem is that pyright's constraint solver is not able to handle "conditional constraints". In your example, the second argument to the call `v1 = f1(v0, 1)` introduces one of two constraints: `T1 is a subtype of int` or `T2 is a subtype of int`. When pyright evaluates this argument, it must pick one of these two constraints. Then it asks the constraint solver to solve all of the type parameters based on recorded constraints. Depending on whether `T1` or `T2` is chosen in the prior step, a solution is possible or not possible.

Fixing this is going to be difficult and may require me to completely redesign the constraint solver. That means it's unlikely to be fixed any time soon. Just want to set expectations.

**Comment by erictraut** [comment]
This appears to have been fixed by other changes that I made in pyright 1.1.358, which was just published.
